### ISSUE TITLE
[bug] $Diff type not treated as object in object spread

### ISSUE THREAD

**Issue by Andarist** [facebook/flow#4571]
Not only that this doesnt work, but also a reported error is quite vague

[Simpler reproduced case](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AQVALygDeiookssAXKAHYCuAtgEZYBOANGaCwIbtaAZxTsAlnQDmiAL4BuZOmygAQoRI9KNeszbtZCxNCwpetACQARMZEgAeXJ1UA+dcV4DaAIhQALCZJeoPJGJqB8tPhEpOQAdPEs3ORatABMskA) - thanks to @asolove 

[Original problematic (simplified) code](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AUQG5YB2KAylgE6EUBCAhgM44C8oA3oqKCgJYC2WBijp8MALlBEArnwBGlADSIAvsnTZQAWUEM6AcxbtO3TFgkByAQ10HzSriiwAPFBKEUeRPUtWIAxrBEQqBMRAAm2tb6hgAUVjZmoAAkACI8kJAAPJEJCniEJORUlPRMoABk7CbYFvHR5qDKAHwAlBIAChSwfDxM2TrRTaDMQxxcAUEooBh0FExhw0ZcXAB0a3UG9svqiZYDtlsO-ILCohIAjAAM19dbqlwUWChSFESSWPCgnd29WDGPDFg0EIw1GxgegiBhBiMzmWDCLWMykRyiAA)

**Comment by asolove** [comment]
So, I don't know why that doesn't work. It seems to me like it should.

But, you can actually make your life easier and [just let Flow figure out the argument type itself](https://flow.org/try/#0PTAEAEDMBsHsHcBQiAuBPADgU1AUQG5YB2KAylgE6EUBCAhgM44C8oA3oqKCgJYC2WBijp8MALlBEArnwBGlADSIAvsnTZQAWUEM6AcxbtO3TFgkByAQ10HzSriiwAPFBKEUeRPUtWIAxrBEQqBMRAAm2tb6hgAUVjZmoABUAJQSAAoUsHw8TAA8kQkAfKDMJRxcAUEooBh0FExhEoXRpUZcXAB03fHR9h3qiZY60XbGDvyCwqISAIwADIuL-apcFFgoUhREkljwoJnZuVgx6wyw0ISl5eOgZxeEMXUNWGEpxsrvvqERIwYxbEcLgkACIhFJIJAQZ8gA) instead of explicitly defining the `$Diff` type that `sendMessage` needs to receive.

**Comment by Andarist** [comment]
@asolove thanks for the answer, this is gonna be a public API of the SDK library, so I would prefer keeping this explicit, even if only for documentation purposes.

Still learning Flow, but in general I find myself wanting to type only types of input output of my public API and let flow figure out all in the middle. I find type casting like this in the middle of the function ugly ;p

**Comment by asolove** [comment]
Just FYI, though I added the type to `parsed`, that isn't actually necessary for this to work. I tried removing it, and the `*` still works fine. The other option, if you want to be explicit for documentation purposes, is to not use `$Diff` but instead just manually list out the object type that's required to be passed in, which I _think_ is just `{text: string}`. That's probably more helpful for someone reading the API docs than sorting through what `$Diff` does.

Either way, I think we can close this issue.

**Comment by Andarist** [comment]
> The other option, if you want to be explicit for documentation purposes, is to not use $Diff but instead just manually list out the object type that's required to be passed in, which I think is just {text: string}.

Actually the code example I gave is overly simplified. I've tried to keep it minimal for clarity and brevity. 

> That's probably more helpful for someone reading the API docs than sorting through what $Diff does.

This is not much of a problem to me. At least yet. Those types wont be public for some time, documenting for our developers (which should know what `$Diff` does) is enough.

> Either way, I think we can close this issue.
Quoting urself - "So, I don't know why that doesn't work. It seems to me like it should.". I think this this indicates that there might be some bug involved here, wouldnt close this prematurely until official confirmation why this is happening.

**Comment by asolove** [comment]
Oh, good point, thanks for reminding me!

To leave it open for that bug, it might help to make the title/description more specific for Flow maintainers combing through issues. 

I rearranged a smaller reproducible example:

```js
type A = {
  foo: number,
  bar: string
};

type B = {
  foo: number
};

let b: $Diff<A, B> = { bar: "thing" };
let a: A = {
  ...b,
  foo: 2
}
```

Which produces this error message:

```
11: let a: A = {               ^ object literal. Expected object instead of
12:   ...b,
         ^ A
```

And I think shows that the problem can be summarized as "$Diff type not treated as object in object spread".

**Comment by Andarist** [comment]
fixed in flow 0.57
